src/DevelMain.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Running your app inside GHCi.
+--
+-- > stack ghci
+--
+-- To start your app, run:
+--
+-- > :l DevelMain
+-- > DevelMain.update
+--
+-- You can also call @DevelMain.shutdown@ to stop the app
+--
+-- There is more information about this approach,
+-- on the wiki: https://github.com/yesodweb/yesod/wiki/ghci
+
+module DevelMain where
+
+import           Prelude
+
+import           Control.Concurrent       (MVar, ThreadId, forkIO, killThread,
+                                           newEmptyMVar, putMVar, takeMVar)
+import           Control.Exception        (finally)
+import           Control.Monad            ((>=>))
+import           Data.IORef               (IORef, newIORef, readIORef,
+                                           writeIORef)
+import           Foreign.Store            (Store (..), lookupStore, readStore,
+                                           storeAction, withStore)
+import           GHC.Word                 (Word32)
+import           Network.Wai.Handler.Warp (defaultSettings, runSettings,
+                                           setPort)
+
+import           Init                     (initialize, shutdownApp)
+
+-- | Start or restart the server.
+-- newStore is from foreign-store.
+-- A Store holds onto some data across ghci reloads
+update :: IO ()
+update = do
+    mtidStore <- lookupStore tidStoreNum
+    case mtidStore of
+      -- no server running
+      Nothing -> do
+          done <- storeAction doneStore newEmptyMVar
+          tid <- start done
+          _ <- storeAction (Store tidStoreNum) (newIORef tid)
+          return ()
+      -- server is already running
+      Just tidStore -> restartAppInNewThread tidStore
+  where
+    doneStore :: Store (MVar ())
+    doneStore = Store 0
+
+    -- shut the server down with killThread and wait for the done signal
+    restartAppInNewThread :: Store (IORef ThreadId) -> IO ()
+    restartAppInNewThread tidStore = modifyStoredIORef tidStore $ \tid -> do
+        killThread tid
+        withStore doneStore takeMVar
+        readStore doneStore >>= start
+
+
+    -- | Start the server in a separate thread.
+    start :: MVar () -- ^ Written to when the thread is killed.
+          -> IO ThreadId
+    start done = do
+        (port, config, app) <- initialize
+        forkIO (finally (runSettings (setPort port defaultSettings) app)
```

Comment:
You probably want to use `forkFinally`. Check the _`mask` and `forkIO`_ section here: https://www.safaribooksonline.com/library/view/parallel-and-concurrent/9781449335939/ch09.html

Reply:
Good call! I'll open an issue on the Yesod template, since it's copied from there :)